### ISSUE TITLE
test() dependencies not added to PATH

### ISSUE THREAD

**Issue by lazka** [mesonbuild/meson#4668]
Hey, support question:

(context: https://gitlab.gnome.org/GNOME/gobject-introspection/blob/master/tests/scanner/annotationparser/meson.build)

We have a test() which calls Python, which uses some Python module and we set PYTHONPATH there so it can find it. The Python module in turn links against various DLLs when the dependencies are build in subprojects they are obviously not in PATH and things fail.

I tried adding the Python module target as a test depends, but that doesn't add anything to PATH. I can add everything manually to PATH, but well...

Is this something which needs fixing in meson or am I missing something?

**Comment by dcbaker** [comment]
The problem is you having a layout like:
dir/mod.py
subproject/foo/foo.dll

and mod.py needs foo.dll?

**Comment by lazka** [comment]
yes

test() -> foo.py -> bar.pyd -> subproject/baz/baz.dll

**Comment by dcbaker** [comment]
I don't believe there is a robust solution for this ATM. if you only need dll's out of one directory you can probably use the `workdir` parameter. This really needs to be resolved though, as I have some use cases for this as well.

**Comment by lazka** [comment]
My current solution is this: https://gitlab.gnome.org/GNOME/gobject-introspection/blob/13e8c7ff80f9ff2dc77e9d0da4b7396182c63612/tests/meson.build#L2

**Comment by dcbaker** [comment]
I wonder if we could use an `environment` object for this?

**Comment by nirbheek** [comment]
We already have the machinery to do this when you pass a build target or a custom target as an argument to a test; we should just do the same when you pass one of those to `depends:`. See: `self.determine_windows_extra_paths()` in `write_test_serialisation()` in `backends.py`.

**Comment by pwithnall** [comment]
We’re [hitting this in GLib too](https://gitlab.gnome.org/GNOME/glib/-/merge_requests/4564#note_2449530), where we have some tests written in Python, which then spawn a native binary in the builddir when they’re run. The native built binary is correctly listed in `depends` on the `test()`. On Linux this works fine as the `write_test_serialisation()` code adds the binary’s dependency paths to `LD_LIBRARY_PATH`. On mingw/Windows it fails, as the same is not done for `PATH`.

For the moment I’ve copied Christoph’s workaround, as fixing and testing Meson issues on Windows is not something I’m set up to do (sorry). Mentioning it here so the use case of running a native built binary from a Python unit test is recorded somewhere, as it’s slightly different from Christoph’s original case of a built Python module.
